docs/testing-guide/contract-testing.md

```diff
@@ -0,0 +1,190 @@
+        self.has_voted[Txn.sender] = UInt64(1)
+        return self.votes.value
+
+    @arc4.abimethod(readonly=True)
+    def get_votes(self) -> UInt64:
+        return self.votes.value
+
+    @arc4.abimethod
+    def change_topic(self, new_topic: UInt64) -> None:
+        assert Txn.sender == Txn.application_id.creator_address(), "Only creator can change topic"
+        self.topic.value = new_topic
+        self.votes.value = UInt64(0)
+        # Reset all votes (this is simplified and not efficient for many users)
+        self.has_voted.delete()
+
+...
+
+@pytest.fixture
+def context():
+    with algopy_testing_context() as ctx:
+        yield ctx
+        ctx.reset()
+
+def test_simple_voting_contract(context):
+    # Arrange
+    initial_topic = UInt64(1)
+    contract = SimpleVotingContract()
+
+    # Act - Create the contract
+    contract.create(initial_topic)
+
+    # Assert - Check initial state
+    assert contract.topic.value == initial_topic
+    assert contract.votes.value == UInt64(0)
+
+    # Act - Vote
+    # note how `algopy.Txn` is temporarily overriden to emulate the transaction content
+    # the method `.vote()` is decorated with `algopy.arc4.abimethod`, which means it will assemble a transaction to emulate the AVM application call
+    with context.scoped_txn_fields(sender=context.default_creator):
+        result = contract.vote()
+
+    # Assert - you can access the corresponding auto generated application call transaction via test context
+    assert len(context.get_transaction_group()) == 1
+
+    # Assert - Note how local and global state are accessed via regular python instance attributes
+    assert result == UInt64(1)
+    assert contract.votes.value == UInt64(1)
+    assert contract.has_voted[context.default_creator] == UInt64(1)
+
+    # Act - Change topic
+    new_topic = UInt64(2)
+    with context.scoped_txn_fields(sender=context.default_creator):
+        contract.change_topic(new_topic)
+
+    # Assert - Check topic changed and votes reset
+    assert contract.topic.value == new_topic
+    assert contract.votes.value == UInt64(0)
+    assert contract.has_voted[context.default_creator] == UInt64(0)
+
+    # Act - Get votes (should be 0 after reset)
+    votes = contract.get_votes()
+
+    # Assert - Check votes
+    assert votes == UInt64(0)
+```
+
+For more examples of tests using `algopy.ARC4Contract`, see the [examples](../examples.md) section.
+
+## `algopy.Contract``
+
+Classes prefixed with `algopy.Contract` (parent class of `algopy.ARC4Contract`) are **required** to be instantiated withing test context. As part of instantiation, the test context will automatically create a matching `algopy.Application` object instance. This behaviour is identical to `algopy.ARC4Contract` class instances.
+
+Unlike `algopy.ARC4Contract`, `algopy.Contract` requires manual setup of the transaction context and explicit method calls. Here's an example demonstrating how to test a `Contract` class:
```

Comment:
This can be reworked a bit now, the active_txn_overrides parameter now allows specifying app_args and foreign arrays, so providing a gtxns is only required if multiple transactions are involved